src/variable_rate.jl

```diff
@@ -0,0 +1,397 @@
+        rootfind = jump.rootfind,
+        interp_points = jump.interp_points,
+        save_positions = jump.save_positions,
+        abstol = jump.abstol,
+        reltol = jump.reltol)
+    return new_cb
+end
+
+function build_variable_callback(cb, idx, jump, jumps...; rng = DEFAULT_RNG)
+    idx += 1
+    new_cb = wrap_jump_in_callback(idx, jump; rng)
+    build_variable_callback(CallbackSet(cb, new_cb), idx, jumps...; rng = DEFAULT_RNG)
+end
+
+function build_variable_callback(cb, idx, jump; rng = DEFAULT_RNG)
+    idx += 1
+    CallbackSet(cb, wrap_jump_in_callback(idx, jump; rng))
+end
+
+@inline function update_jumps!(du, u, p, t, idx, jump)
+    idx += 1
+    du[idx] = jump.rate(u.u, p, t)
+end
+
+@inline function update_jumps!(du, u, p, t, idx, jump, jumps...)
+    idx += 1
+    du[idx] = jump.rate(u.u, p, t)
+    update_jumps!(du, u, p, t, idx, jumps...)
+end
+
+################################### VR_Direct ####################################
+
+"""
+$(TYPEDEF)
+
+A concrete `VariableRateAggregator` implementing a direct method-based approach for
+simulating `VariableRateJump`s. `VR_Direct` (Variable Rate Direct Callback) efficiently
+samples jump times using one continuous callback to integrate the total intensity /
+propensity for all `VariableRateJump`s, sample when the next jump occurs, and then sample
+which jump occurs at this time. 
+
+## Examples
+Simulating a birth-death process with `VR_Direct` (default):
+```julia
+using JumpProcesses, OrdinaryDiffEq  
+u0 = [1.0]           # Initial population  
+p = [10.0, 0.5]      # [birth rate, death rate coefficient]  
+tspan = (0.0, 10.0)  
+
+# Birth jump: ∅ → X  
+birth_rate(u, p, t) = p[1] 
+birth_affect!(integrator) = (integrator.u[1] += 1; nothing)  
+birth_jump = VariableRateJump(birth_rate, birth_affect!)  
+
+# Death jump: X → ∅  
+death_rate(u, p, t) = p[2] * u[1]
+death_affect!(integrator) = (integrator.u[1] -= 1; nothing)  
+death_jump = VariableRateJump(death_rate, death_affect!)  
+
+# Problem setup  
+oprob = ODEProblem((du, u, p, t) -> du .= 0, u0, tspan, p)  
+jprob = JumpProblem(oprob, birth_jump, death_jump; vr_aggregator = VR_Direct())  
+sol = solve(jprob, Tsit5()) 
+```
+
+## Notes  
+- `VR_Direct` is expected to generally be more performant than `VR_FRM`.
+"""
+struct VR_Direct <: VariableRateAggregator end
+
+mutable struct VR_DirectEventCache{T, RNG <: AbstractRNG}
+    prev_time::T
+    prev_threshold::T
+    current_time::T
+    current_threshold::T
+    total_rate_cache::T
+    rng::RNG
+    variable_jumps::Tuple{Vararg{VariableRateJump}}
+    cur_rates::Vector{T}
+
+    function VR_DirectEventCache(jumps::JumpSet, ::Type{T}; rng = DEFAULT_RNG) where T
+        initial_threshold = randexp(rng, T)
+        vjumps = jumps.variable_jumps
+        cur_rates = Vector{T}(undef, length(vjumps))
+        new{T, typeof(rng)}(zero(T), initial_threshold, zero(T), initial_threshold,
+                           zero(T), rng, vjumps, cur_rates)
```

Comment:
@ChrisRackauckas don't we need a call back initialization function that does this, so that if you call `solve` multiple times these fields get reset?